### PR TITLE
Add send_transaction2 support

### PR DESCRIPTION
**Pull request by NatPDeveloper**
This is one of my first open source contributions. I'm very open to feedback.

## Change Description

Support Issue #5 
Related: https://github.com/eosnetworkfoundation/product/pull/30
- Added support for `/v1/chain/send_transaction2`.
- Added rpc unit test for `/v1/chain/send_transaction2`.
- Added `text-encoding` as dev dependency in package.json to run unit tests. 
- Update api.transaction to include options for retrying transaction
- Add error if `server_version_string` below `v3.1.0`
- Update `GetInfoResult` type with latest returned results
- Added `retry transaction` for `node.test.ts`

## API Changes
- added `returnFailureTrace` option, default false
- added `retryIrreversible` option, default false
- added `retryTrxNumBlocks` option, default 0
- added `useOldRPC` option, default false
- added `useOldSendRPC` option, default false

## Documentation Additions
- added documentation to the `README.md` file for 5 new API parameters.

### CODE REVIEW

**Comment by ericpassmore** [comment]
Hi Nat, this is good thank you for submitting. I was actually working on the same based on your medium post, and not as far along. I am taking a look now.

**Comment by ericpassmore** [comment]
one ask, can we better align with cleos send_transaction calling patterns? Specifically my ask is to include useOldSendRPC flag to transact. We can add a retry_irreversible later. Suggest patch included below. 

diff --git a/src/eosjs-api.ts b/src/eosjs-api.ts
index aad1e84..c4011be 100644
--- a/src/eosjs-api.ts
+++ b/src/eosjs-api.ts
@@ -223,6 +223,7 @@ export class Api {
      * Named Parameters:
      *    * `broadcast`: broadcast this transaction?
      *    * `sign`: sign this transaction?
+     *    * `useOldSendRPC`: Use old RPC /v1/chain/send_transaction, rather than new RPC /v1/chain/send_transaction2
      *    * `retryTrx`: retry this transaction?
      *    * `returnFailureTrace`: return failure trace for this transaction?
      *    * `retryTrxNumBlocks`: retry trx for X blocks, 0 for irreversible?
@@ -237,6 +238,7 @@ export class Api {
             blocksBehind, 
             expireSeconds,
             returnFailureTrace = false,
+            useOldSendRPC = false,
             retryTrx = false,
             retryTrxNumBlocks = 30
         }:
@@ -246,6 +248,7 @@ export class Api {
             blocksBehind?: number; 
             expireSeconds?: number; 
             returnFailureTrace?: boolean;
+            useOldSendRPC?: boolean;
             retryTrx?: boolean;
             retryTrxNumBlocks?: number;
         } = {}
@@ -257,10 +260,19 @@ export class Api {
             this.chainId = info.chain_id;
         }
 
-        if(retryTrx && info.server_version_string < "v3.1.0") {
-            throw new Error(`Retry transaction feature unavailable for nodeos :${info.server_version_string}`);
+        /**
+         * Safety checks to make sure retryTrx can be used
+         */
+        if(retryTrx) {
+            if (info.server_version_string < "v3.1.0") {
+                throw new Error(`Retry transaction feature unavailable for nodeos :${info.server_version_string}`);
+            }
+            if (useOldSendRPC) {
+                throw new Error('Retry transaction feature not compatible with useOldSendRPC');
+            }
         }
 
+
         if (typeof blocksBehind === 'number' && expireSeconds) { // use config fields to generate TAPOS if they exist
             if (!info) {
                 info = await this.rpc.get_info();
@@ -268,6 +280,12 @@ export class Api {
             const refBlock = await this.rpc.get_block_header_state(info.head_block_num - blocksBehind);
             transaction = { ...ser.transactionHeader(refBlock, expireSeconds), ...transaction };
         }
+        /**
+         * TODO can add retry_irreversible logic here
+         *  Request node to retry transaction until it is irreversible or expires, blocking call
+         *  // Set tapos, default to last irreversible block if it's not specified by the user
+         *  refBlock = info.last_irreversible_block_id;
+         */
 
         if (!this.hasRequiredTaposFields(transaction)) {
             throw new Error('Required configuration or TAPOS fields are not present');
@@ -297,7 +315,9 @@ export class Api {
             });
         }
         if (broadcast) {
-            if(retryTrx) {
+            if (useOldSendRPC) {
+                return this.pushSignedTransaction(pushTransactionArgs);
+            } else {
                 return this.sendSignedTransaction2({
                     return_failure_trace: returnFailureTrace, 
                     retry_trx: retryTrx, 
@@ -305,7 +325,7 @@ export class Api {
                     transaction: pushTransactionArgs
                 });
             }
-            return this.pushSignedTransaction(pushTransactionArgs);
+
         }
         return pushTransactionArgs;
     }
@@ -324,14 +344,14 @@ export class Api {
     /** Broadcast a signed transaction2 */
     public async sendSignedTransaction2(
         { 
-            return_failure_trace, 
+            return_failure_trace,
             retry_trx, 
             retry_trx_num_blocks, 
             transaction: { signatures, serializedTransaction, serializedContextFreeData } 
         }: SendTransaction2Args
     ): Promise<any> {
         return this.rpc.send_transaction2({
-            return_failure_trace, 
+            return_failure_trace,
             retry_trx, 
             retry_trx_num_blocks, 
             transaction: {

**Comment by NatPDeveloper** [comment]
Hm looks like the wires got crossed there. Perhaps you can do a diff check with https://www.diffchecker.com/ and share the link for the various files you see being changed.

Yes I agree it should match cleos. In the meantime I'll try updating what parts I can piece together.

**Comment by ericpassmore** [comment]
Here is the diff 
https://www.diffchecker.com/9MErjwI0

**Comment by NatPDeveloper** [comment]
Okay great, I added that as well as the irreversible feature, let me know what your thoughts are.

I also update the PR notes at the top to reflect removing the `retryTrx` option to match up with cleos.

**Comment by ericpassmore** [comment]
Exciting stuff, thanks for the quick turnaround. I did a quick check and it looks good. I want to go back and closer look at irreversible updates in cleos code for comparison.

**Comment by NatPDeveloper** [comment]
I've added a change to not use the retry feature by default and to return the failure trace by default to also match cleos.

```
cleos --print-request -u http://$NODEOS push action eosio.token transfer '["heliosselene","uuddlrlrbass","0.0001 EOS",""]' -p heliosselene
...
POST /v1/chain/send_transaction2 HTTP/1.0
Host: 3.91.147.242:8888
content-length: 467
Accept: */*
Connection: close

{
  "return_failure_trace": true,
  "retry_trx": false,
  "transaction": {
    "signatures": [
      "SIG_K1_KAH8gQeK74FP3RU53C5bNrFk6DZGRzeHMPX4bJnJFZaiRpgMRvzerEfTCFXdDbWQN2sTnbp2GyTJAz3qocyFW5T3m2Qvt6"
    ],
    "compression": "none",
    "packed_context_free_data": "",
    "packed_trx": "47e6ef6216849415f736000000000100a6823403ea3055000000572d3ccdcd01a0a68a0a63eaa26a00000000a8ed323221a0a68a0a63eaa26a80b13937de9892d6010000000000000004454f53000000000000"
  }
}
---------------------
executed transaction: 21ae7ccc96b342a61cd151cedd75281922416b632d6d4ef1ca33bbf3a5f4721c  128 bytes  229 us
#   eosio.token <= eosio.token::transfer        {"from":"heliosselene","to":"uuddlrlrbass","quantity":"0.0001 EOS","memo":""}
#  heliosselene <= eosio.token::transfer        {"from":"heliosselene","to":"uuddlrlrbass","quantity":"0.0001 EOS","memo":""}
#  uuddlrlrbass <= eosio.token::transfer        {"from":"heliosselene","to":"uuddlrlrbass","quantity":"0.0001 EOS","memo":""}
warning: transaction executed locally, but may not be confirmed by the network yet         ] 
```

File: src/eosjs-api.ts [hunk 1]
```diff
@@ -273,12 +313,35 @@ export class Api {
...
-            return this.pushSignedTransaction(pushTransactionArgs);
+            if(isRetry) {
+                let params: SendTransaction2Args = {
+                    return_failure_trace: returnFailureTrace, 
+                    retry_trx: true,
+                    transaction: pushTransactionArgs
+                }
+                if(retryTrxNumBlocks > 0) {
+                    params = {
+                        ...params,
```
Comment (ericpassmore) [inline src/eosjs-api.ts:324]: irreversible well done

**Review by ericpassmore — APPROVED** [review]

**Review by ericpassmore — APPROVED** [review]

**Comment by ericpassmore** [comment]
updates for eosjs, pr has good change desc

**Pull request merged**